jetty-start/src/test/java/org/eclipse/jetty/start/CommandLineBuilderTest.java

```diff
@@ -23,43 +28,87 @@ public class CommandLineBuilderTest
+        assertThat(cmd.toCommandLine(), is("java -Djava.io.tmpdir='/home/java/temp dir/' --version"));
     }
 
     @Test
     public void testSimpleHomeNoSpace()
     {
-        CommandLineBuilder cmd = new CommandLineBuilder("java");
-        cmd.addEqualsArg("-Djetty.home", "/opt/jetty");
-        assertThat(cmd.toQuotedString(), is("java -Djetty.home=/opt/jetty"));
+        CommandLineBuilder cmd = new CommandLineBuilder();
+        cmd.addArg("java");
+        cmd.addArg("-Djetty.home", "/opt/jetty");
+        assertThat(cmd.toCommandLine(), is("java -Djetty.home=/opt/jetty"));
     }
 
     @Test
     public void testSimpleHomeWithSpace()
     {
-        CommandLineBuilder cmd = new CommandLineBuilder("java");
-        cmd.addEqualsArg("-Djetty.home", "/opt/jetty 10/home");
-        assertThat(cmd.toQuotedString(), is("java '-Djetty.home=/opt/jetty 10/home'"));
+        CommandLineBuilder cmd = new CommandLineBuilder();
+        cmd.addArg("java");
+        cmd.addArg("-Djetty.home", "/opt/jetty 10/home");
+        assertThat(cmd.toCommandLine(), is("java -Djetty.home='/opt/jetty 10/home'"));
     }
 
     @Test
     public void testEscapedFormattingString()
     {
-        CommandLineBuilder cmd = new CommandLineBuilder("java");
-        cmd.addEqualsArg("-Djetty.home", "/opt/jetty");
-        cmd.addEqualsArg("jetty.requestlog.formatter", "%{client}a - %u %{dd/MMM/yyyy:HH:mm:ss ZZZ|GMT}t \"%r\" %s %O \"%{Referer}i\" \"%{User-Agent}i\"");
-        assertThat(cmd.toQuotedString(), is("java -Djetty.home=/opt/jetty 'jetty.requestlog.formatter=%{client}a - %u %{dd/MMM/yyyy:HH:mm:ss ZZZ|GMT}t \"%r\" %s %O \"%{Referer}i\" \"%{User-Agent}i\"'"));
+        CommandLineBuilder cmd = new CommandLineBuilder();
+        cmd.addArg("java");
+        cmd.addArg("-Djetty.home", "/opt/jetty");
+        cmd.addArg("jetty.requestlog.formatter", "%{client}a - %u %{dd/MMM/yyyy:HH:mm:ss ZZZ|GMT}t \"%r\" %s %O \"%{Referer}i\" \"%{User-Agent}i\"");
+        assertThat(cmd.toCommandLine(), is("java -Djetty.home=/opt/jetty jetty.requestlog.formatter='%{client}a - %u %{dd/MMM/yyyy:HH:mm:ss ZZZ|GMT}t \"%r\" %s %O \"%{Referer}i\" \"%{User-Agent}i\"'"));
     }
 
     @Test
     public void testEscapeUnicode()
     {
-        CommandLineBuilder cmd = new CommandLineBuilder("java");
-        cmd.addEqualsArg("-Djetty.home", "/opt/jetty");
-        cmd.addEqualsArg("monetary.symbol", "€");
-        assertThat(cmd.toQuotedString(), is("java -Djetty.home=/opt/jetty monetary.symbol=€"));
+        CommandLineBuilder cmd = new CommandLineBuilder();
+        cmd.addArg("java");
+        cmd.addArg("-Djetty.home", "/opt/jetty");
+        cmd.addArg("monetary.symbol", "€");
+        assertThat(cmd.toCommandLine(), is("java -Djetty.home=/opt/jetty monetary.symbol='€'"));
+    }
+
+    public static Stream<Arguments> shellQuoting()
+    {
+        return Stream.of(
+            Arguments.of(null, null),
+            Arguments.of("", "''"),
```

Comment:
This is weird, why it is not the empty string `""` but has single quotes?

Reply:
because somehow somebody gave us an argument that was the empty string.  If we just return the empty string then we will lose the argument.  Returning empty quotes preserves that there in an argument of zero length

Reply:
@sbordet this is particularly important when the argument has secondary arguments.

Like the `--classpath` argument on Java itself, that argument alone isn't useful, it needs the next argument to be valid.
 `--classpath <value>` would be two arguments (1. to indicate the desire to use classpath, and 2. to indicate the value of classpath).

So the use of `--classpath ''` would be two arguments, a classpath that is empty (I know, not terribly useful for Java, but it shows an example of an empty argument)

In jetty `start.jar` terms, this particular example (`--classpath ''`) shows up when you use `start.jar --dry-run` in a directory that has no libs defined (eg: an empty directory, or no directly declared libs, or no enabled modules that have libs)